Sources/PenguinStructures/NumberOperations.swift

```diff
@@ -0,0 +1,72 @@
+// Copyright 2020 Penguin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extension BinaryInteger {
+  /// Returns a sequence of the positive integers that are co-prime with `self`.
+  ///
+  /// Definition: Two numbers are co-prime if their GCD is 1.
+  public var positiveCoprimes: PositiveCoprimes<Self> { PositiveCoprimes(self) }
+}
+
+/// A sequence of numbers that are co-prime with `n`, up to `n`.
+public struct PositiveCoprimes<Number: BinaryInteger>: Sequence {
```

Comment:
I'd rather instead of making it conform to collection instead make it a sequence. Concretely, we know there's infinitely many primes, and there's no reason why we should stop at the number itself, so I propose that this actually be an (infinite) sequence instead.

I also dislike the collection protocol here because the size of the collection is unknown until it's been computed, which reduces the value of it being a constant storage abstraction.

I like `Domain`, but I dislike `limit`, and think `n` (or perhaps `b`) is more appropriate appropriate. 

Edit: I'm fine with `target` (as you suggested), and have made the corresponding edits, but I'm still not fully convinced this is the right choice...

Reply:
Even if it was infinite, Collection is the right abstraction.  Please see “But sequences can be infinite” at https://forums.swift.org/t/pitch-make-collection-super-convenient-and-retire-sequence/12103

The fundamental difference between Collection and Sequence is single vs multipass, and all that implies in terms of algorithms: the ability to subdivide it, to binary search, etc.

I don't know what you're saying about the `count` and the value of it being a constant storage abstraction.  There are *many* examples of collections whose `count` is an O(n) computation.  I lost the battle to make count be a method because of this long ago.


The one argument I would consider worthwhile against `Collection` conformance is that the cost of traversal from one element to the next is probably not O(1).  But even that is IMO not a good enough reason to avoid collection conformance; what you do is note in docs that it doesn't provide expected performance, as lazy filter collection does.  In general, it's more valuable to have the functionality available even if it's going to be slow.

I'm not at all in love with `target` either.  It's weird that it plays two roles: both the limit and a coprime of the all the elements, which makes it hard to name.  `n` or `b` would make sense for function parameter names, or if that was a standard name used for a coprime in math, like you use *m* and *b* for a linear equation *y*=*mx*+*b*. Otherwise, a stored property has a persistent meaning that is not available from the doc comment near its use site and is thus usually worth naming more explicitly.

Reply:
From the first post in https://forums.swift.org/t/pitch-make-collection-super-convenient-and-retire-sequence/12103:

> A true stream would require significant additional storage or computation to
support multiple passes over its elements. [...]
> The initial state of the element generation procedure, which is modified as new elements are generated, is very costly to construct and to store, such as in the Mersenne Twister 4 pseudo-random number generator.
> [...]

I would argue that this type falls into roughly the same category. This algorithm seems to me to fall into the same category. Concretely: while it's possible to store previous values (a-la Mersenne Twister) to facilitate indexing, or do an increasing amount of re-computation, these are both unsatisfactory.

Reply:
You don't need to store previous values.  The index just needs to store the same thing as the Iterator does.  https://gist.github.com/dabrahams/6119b718aab45d8ef59e141eb37c711e

Reply:
Hmmm, by switching to `Collection`, I've gotten rid of all mutable stored properties, and also got rid of the need to create a nested `Iterator` or `Index` type (by re-using `Domain`). While it demands a bit more code, and I'm pretty sure operations like `prefix` are probably somewhat inefficient, this does seem like an interesting design. It also makes certain functionality pretty easy, such as finding the next positive coprime greater than any arbitrary number.

Reply:
> operations like prefix are probably somewhat inefficient

You could always define a lazy `SubSequence` type to get around that.

> It also makes certain functionality pretty easy, such as finding the next positive coprime greater than any arbitrary number.

Exactly.  That's why, “if it *can* conform, it *should* conform.”

Reply:
> You could always define a lazy `SubSequence` type to get around that.

I wondered about that, but the implementation of [`prefix(_:)`](https://github.com/apple/swift/blob/acb16cf216e37010f8e372fc4fa38964a0364c83/stdlib/public/core/Collection.swift#L1311) doesn't appear to admit an implementation where the subsequence is lazy, as it relies upon directly computing the `end` index from the call to `index(_:offsetBy:limitedBy:)` which can't be lazy, and IIUC, there's no `.lazy.prefix(n)` (although there is a `LazyPrefixWhileSequence`... haven't played in enough contexts to fully grok what would happen in all situations).